### PR TITLE
Split a func into cold/hot parts, reducing binary size

### PR DESCRIPTION
**Pull request by sivadeilra**
I noticed that the Size::bits function is called in many places,
and is inlined into them. On x86_64-pc-windows-msvc, this function
is inlined 527 times, and compiled separately (non-inlined) 3 times.

Each of those inlined calls contains code that panics. This commit
moves the `panic!` call into a separate function and marks that
function with `#[cold]`.

This reduces binary size by 24 KB. Not much, but it's something.
Changes like this often reduce pressure on instruction-caches,
since it reduces the amount of code that is inlined into hot code
paths. Or more precisely, it removes cold code from hot cache lines.

### CODE REVIEW

**Comment by camelid** [comment]
Does this reduce the size of the `rustc` binary only or every binary?

**Comment by sivadeilra** [comment]
> Does this reduce the size of the `rustc` binary only or every binary?

It reduces the size of the `rustc_driver-xxx.dll` binary. If you're asking about the intermediate `.rlib` files, then I assume it is going to reduce the size of the ones affected (`rustc_target`, `rustc_middle`, `rustc_mir`, `rustc_codegen_llvm`, mainly.) But I'm more interested in the size of the final output than in the size of the intermediate files.

**Comment by bjorn3** [comment]
The difference seems to be within noise.

**Comment by sivadeilra** [comment]
I'm not expecting a big result from this. But eliminating 500 duplicates of cold panic code seems worth it, right?

**Comment by oli-obk** [comment]
> But eliminating 500 duplicates of cold panic code seems worth it, right?

Indeed. Do you think we'll get the same benefit even when keeping the original `checked_mul` and `unwrap_or_else` logic and just using the new function within the unwrap closure?

**Comment by sivadeilra** [comment]
> > But eliminating 500 duplicates of cold panic code seems worth it, right?
> 
> Indeed. Do you think we'll get the same benefit even when keeping the original `checked_mul` and `unwrap_or_else` logic and just using the new function within the unwrap closure?

Possibly. I can check the generated assembly. I've seen `checked_mul` generate fairly efficient code, so I'm not worried about that part. Mainly it's the panicking and formatting that can be surprisingly expensive. Not just to execute, but just the code itself can be surprisingly large, which can be painful when it's duplicated many times.

Ideally, we would notice that a particular basic block always leads to a `#[cold]` path, and would "outline" that BB (prevent it from being inlined, and so prevent it from being dup'd). That would solve the general case. However, that's not something that rustc does automatically at this point. If it is ever added, then a lot of hand-tuning like this could be removed.

**Comment by sivadeilra** [comment]
I tried switching the code back to using `checked_mul`. The generated assembly code is a bit worse; it uses a `mul` instruction, rather than shifting, and a `jo` to branch on overflow.

Maybe a better approach would be for the `Size` struct to contain a `u64` that counts the number of bits, not the number of bytes, in a size. That way, the `bytes` function would just be `fn bytes(self) -> u64 { self.bits >> 3 }`. No conditionals at all. The constructor (`Size::new`) would test for overflow. Which would basically never happen -- we'll never have a legit type whose `Size` is 0x3fff_ffff_ffff_ffff in bytes, that is, a size that would overflow when expressed as a bit-count but not overflow when expressed as a byte-count.

**Comment by oli-obk** [comment]
Hmm... it seems a bit wrong to store the bits if we can't actually encode the bit size, but must always encode a byte size.

One thing we could try is to make the `Size` struct use `rustc_layout_scalar_valid_range_end` to declare that the maximum value must be smaller than `u64::max_value() >> 3` and see whether llvm can figure out that taking the `raw` field will thus allow us to multiply by 8 without causing any problems. This would also have the fun side effect of allowing the free bits of `Size` to be used in enum optimizations across the compiler. Additionally, we now need to check that the size is smaller than `u64::max_value() >> 3` at `Size` construction time. Not sure if that gains us anything to checking it at the use site.

What do you think of that idea? We can also open a compiler MCP (major change proposal) to make sure that we get the input of the team on this, since it *technically* could be observed by users, but I don't think in any case in which some code currently compiles and won't compile after this change.

**Comment by bjorn3** [comment]
> One thing we could try is to make the Size struct use rustc_layout_scalar_valid_range_end to declare that the maximum value must be smaller than u64::max_value() >> 3 and see whether llvm can figure out that taking the raw field will thus allow us to multiply by 8 without causing any problems.

As far as I know LLVM is not told about `rustc_layout_scalar_valid_range_end`.

**Comment by sivadeilra** [comment]
I think using `rustc_layout_scalar_valid_range_end` is interesting. `Option<Size>` is used in a few places. Not enough that I think the benefits are huge, but it's always good to enable more layout optimizations.

I think the key thing is to move the validation to the constructor of `Size`. Everything else is just "how do we represent this". It seems very unlikely that any code relies on constructing a very large `Size` and then carefully avoids calling `my_size.bits()` on it; the calls to `bits()` greatly outnumber the calls to `bytes()`.

Me personally, I think this is below the threshold of the MCP process; the compiler team has much bigger issues facing them. If we make this change (verifying the lower maximum value for `Size::from_*`), and it turns out that something *does* depend on it, then we'll get a deterministic failure. That seems to fit with the philosophy of the nightly / beta release process, rather than the MCP process. We can commit it to `master`, and let it go through the usual stabilization process. If it causes problems, which I think is very unlikely, then it will be easy to undo this change, since it affects only the internals of a type, and not its publicly-visible API.

**Comment by oli-obk** [comment]
I agree on all points. Do you want to implement this as a replacement for this PR or do you have other steps in mind?

**Comment by sivadeilra** [comment]
@oli-obk I think this is ready now.

**Comment by oli-obk** [comment]
Thanks!

@bors r+ rollup

File: compiler/rustc_target/src/abi/mod.rs [hunk 1]
```diff
@@ -238,22 +238,38 @@ pub enum Endian {
 #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Encodable, Decodable)]
 #[derive(HashStable_Generic)]
 pub struct Size {
+    // The top 3 bits are ALWAYS zero.
     raw: u64,
```
Comment (RalfJung) [inline compiler/rustc_target/src/abi/mod.rs:242]: I am concerned by this comment that is being added here.
- It doesn't seem actually enforced -- with `Size::from_bytes` I can easily construct a `Size` that violates this.
- I would like to use this type to represent the size of any possible Rust object in Miri (including for things like `size_of_val_raw`, where the object does not actually have to exist in the address space). However, the size limit for that is `isize::MAX`. Something seems off if the compiler's `Size` type cannot be used to represent the size of all objects in the language...

Currently Miri enforces `dl.obj_size_bound()` as the size limit, but that is probably wrong -- it does not match what the [reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html) and [`size_of_val_raw`](https://doc.rust-lang.org/std/mem/fn.size_of_val_raw.html) say the limit should be (they both say it is `isize::MAX`).

File: compiler/rustc_target/src/abi/mod.rs [hunk 2]
```diff
@@ -238,22 +238,38 @@ pub enum Endian {
...
+
+        #[cold]
+        fn overflow(bits: u64) -> ! {
+            panic!("Size::from_bits({}) has overflowed", bits);
+        }
+
+        // This is the largest value of `bits` that does not cause overflow
+        // during rounding, and guarantees that the resulting number of bytes
+        // cannot cause overflow when multiplied by 8.
+        if bits > 0xffff_ffff_ffff_fff8 {
```
Comment (RalfJung) [inline compiler/rustc_target/src/abi/mod.rs:261]: I am also very confused by this check. `bits / 8 + ((bits % 8) + 7) / 8` will *never* overflow I think, so what does this have to do with "overflow during rounding"?
